cg/apps/demultiplex/demultiplex_api.py

```diff
@@ -0,0 +1,112 @@
+    """
+    for child in in_path.iterdir():
+        if not child.is_dir():
+            continue
+        LOG.info("Found directory %s", child)
+        yield child
+
+
+class DemultiplexingAPI:
+    """Demultiplexing API should deal with anything related to demultiplexing
+
+    This includes starting demultiplexing, creating sample sheets, creating base masks
+    """
+
+    def __init__(self, config: dict, out_dir: Optional[Path] = None):
+        self.slurm_api = SlurmAPI()
+        self.slurm_account: str = config["demultiplex"]["slurm"]["account"]
+        self.mail: str = config["demultiplex"]["slurm"]["mail_user"]
+        self.out_dir: Path = out_dir or Path(config["demultiplex"]["out_dir"])
+        self.dry_run: bool = False
+
+    def set_dry_run(self, dry_run: bool) -> None:
+        LOG.info("Set dry run to %s", dry_run)
+        self.dry_run = dry_run
+        self.slurm_api.set_dry_run(dry_run=dry_run)
+
+    @staticmethod
+    def get_sbatch_error(flowcell_id: str, log_path: Path, email: str) -> str:
+        """Create the sbatch error string"""
+        error_parameters: SbatchError = SbatchError(
+            flowcell_name=flowcell_id, email=email, logfile=log_path.as_posix()
+        )
+        return DEMULTIPLEX_ERROR.format(**error_parameters.dict())
+
+    @staticmethod
+    def get_sbatch_command(run_dir: Path, out_dir: Path, sample_sheet: Path) -> str:
+        command_parameters: SbatchCommand = SbatchCommand(
+            run_dir=run_dir.as_posix(),
+            out_dir=out_dir.as_posix(),
+            sample_sheet=sample_sheet.as_posix(),
+        )
+        return DEMULTIPLEX_COMMAND.format(**command_parameters.dict())
+
+    @staticmethod
+    def get_demultiplex_sbatch_path(directory: Path) -> Path:
+        """Get the path to where sbatch file should be kept"""
+        return directory / "demux-novaseq.sh"
+
+    @staticmethod
+    def get_logfile(out_dir: Path, flowcell: Flowcell) -> Path:
+        """Create the path to the logfile"""
+        # How to handle if a flowcell is redemuxed ?
+        return out_dir / f"project.{flowcell.flowcell_id}.log"
+
+    def start_demultiplexing(self, flowcell: Flowcell):
+        """Start demultiplexing for a flowcell"""
+        flowcell_out_dir: Path = self.out_dir / flowcell.path.name
+        LOG.info("Demultiplexing to %s", flowcell_out_dir)
+        if not self.dry_run:
+            LOG.info("Creating out dir %s", flowcell_out_dir)
+            flowcell_out_dir.mkdir(exist_ok=False, parents=True)
+        log_path: Path = self.get_logfile(out_dir=flowcell_out_dir, flowcell=flowcell)
+        error_function: str = self.get_sbatch_error(
+            flowcell_id=flowcell.flowcell_id, log_path=log_path, email=self.mail
+        )
+        commands: str = self.get_sbatch_command(
+            run_dir=flowcell.path, out_dir=flowcell_out_dir, sample_sheet=flowcell.sample_sheet_path
+        )
+
+        sbatch_info = {
```

Comment:
create Sbatch object directly instead of making dict and parsing? Dict is not reused anyway

Reply:
changed